src/axislike.rs

```diff
@@ -106,3 +112,236 @@ impl AxisPair {
+        FloatOrd(self.x_positive_low).hash(state);
+        FloatOrd(self.x_negative_low).hash(state);
+        FloatOrd(self.y_positive_low).hash(state);
+        FloatOrd(self.y_negative_low).hash(state);
+    }
+}
+
+#[allow(clippy::doc_markdown)] // False alarm because it thinks DPad is an un-quoted item
+/// A virtual DPad that you can get an [`AxisPair`] from
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct VirtualDPad {
+    /// The input that represents the up direction in this virtual DPad
+    pub up: InputKind,
+    /// The input that represents the down direction in this virtual DPad
+    pub down: InputKind,
+    /// The input that represents the left direction in this virtual DPad
+    pub left: InputKind,
+    /// The input that represents the right direction in this virtual DPad
+    pub right: InputKind,
+}
+
+impl VirtualDPad {
+    /// Generates a [`VirtualDPad`] corresponding to the arrow keyboard keycodes
+    pub const fn arrow_keys() -> VirtualDPad {
+        VirtualDPad {
+            up: InputKind::Keyboard(KeyCode::Up),
+            down: InputKind::Keyboard(KeyCode::Down),
+            left: InputKind::Keyboard(KeyCode::Left),
+            right: InputKind::Keyboard(KeyCode::Right),
+        }
+    }
+
+    /// Generates a [`VirtualDPad`] corresponding to the `WASD` keyboard keycodes
+    pub const fn wasd() -> VirtualDPad {
+        VirtualDPad {
+            up: InputKind::Keyboard(KeyCode::W),
+            down: InputKind::Keyboard(KeyCode::S),
+            left: InputKind::Keyboard(KeyCode::A),
+            right: InputKind::Keyboard(KeyCode::D),
+        }
+    }
+
+    /// Generates a [`VirtualDPad`] corresponding to the DPad on a gamepad
+    pub const fn dpad() -> VirtualDPad {
+        VirtualDPad {
+            up: InputKind::GamepadButton(GamepadButtonType::DPadUp),
+            down: InputKind::GamepadButton(GamepadButtonType::DPadDown),
+            left: InputKind::GamepadButton(GamepadButtonType::DPadLeft),
+            right: InputKind::GamepadButton(GamepadButtonType::DPadRight),
+        }
+    }
+
+    /// Generates a [`VirtualDPad`] corresponding to the face buttons on a gamepad
+    ///
+    /// North corresponds to up, west corresponds to left, east corresponds to right, south corresponds to down
+    pub const fn gamepad_face_buttons() -> VirtualDPad {
+        VirtualDPad {
+            up: InputKind::GamepadButton(GamepadButtonType::North),
+            down: InputKind::GamepadButton(GamepadButtonType::South),
+            left: InputKind::GamepadButton(GamepadButtonType::West),
+            right: InputKind::GamepadButton(GamepadButtonType::East),
+        }
+    }
+
+    /// Constructs a [`VirtualDpad`] corresponding to the left analogue stick of a gamepad
+    ///
+    /// A small, symettric deadzone is added for convenience.
+    /// If you wish to control the dead zone yourself, construct this struct manually.
+    pub const fn left_stick() -> VirtualDPad {
```

Comment:
I love the new `arrow_keys`, `wasd`, `dpad`, and `gamepad_face_buttons` helpers, but I'm a little confused about the `left_stick`, `right_stick`, and `from_dual_gamepad_axis` helpers.

Why would you use a `VirtualDPad` with a dual gamepad axis inside it instead of just using a `DualGamepadAxis`?

It seems like it might be confusing to a user who notices that they could either use `VirtualDPad::left_stick()` or `DualGamepadAxis::left_stick()`.

The only practical difference I can think of is that a `DualGamepadAxis` can be put in a `Chord` or combined with other axes in a `VirtualDPad`, while a `VirtualDPad` can't be used in chords or combined with other inputs.

Reply:
Oh I see! This is my fault for not quite understanding the API correctly. Let me fix that and leave a clarifying doc comment.